batchflow/notifier.py

```diff
@@ -248,11 +266,37 @@ def update_plots(self, add_suptitle=False):
         if add_suptitle:
             title = self.format_meter(self.n, self.total, time()-self.start_t, ncols=80)
             plt.suptitle(title, y=0.99, fontsize=14)
+
+        if savepath:
+            plt.savefig(savepath, bbox_inches='tight', pad_inches=0)
         plt.show()
 
     # Convenient alias for working inspect instance after pipeline run
     visualize = update_plots
 
+    def update_file(self):
+        with open(self.file, 'a+') as f:
+            timestamp = strftime("%Y-%m-%d  %H:%M:%S", self.timestamps[-1])
+            msg = f'Iteration {self.bar.n:5};    {self.bar.desc[:-1]}'
+            print(f'{timestamp}     {msg}', file=f)
+
+    def to_file(self, file):
```

Comment:
This method doesn't reuse any code that prints logs on the fly. This can lead to differences between reports that are generated online and those printed upon request.

It might be more convenient to use `logging` standard library. It allows to define consistent log format, including timestamps. One can also choose different handlers to control the destination of log output. 

writing to a `logger.FileHandler` in `update_file` would be an equivalent of current `to_file`, and using a stream and flushing it would produce online logging.

IMO, there is no need to have a logger that prints information strictly on each update and this can be also very slow. Having a FileHandler that controls output to a file would be sufficient.

Reply:
I am well aware of the `logging` module, yet it requires some care to use.
The reason why I chose not to use it here is because our `FileLogger` in `Research` collects all the info from all the loggers and puts it into `research.log` file. Furious spam from the iteration-wise notifiers from multiple processes would make this file completely unreadable; that is why I decided to use simple file writes, provided we don't need any extra functionality (for now).

Reply:
As for the not reusing code: it hurts me to say it, but `premature generalization is the root of all evil`, as one man (almost) said. The effort required to generalize similar 3 lines of code in different places may just make it harder to read and lead to more confusion, not clarity.

Reply:
> I am well aware of the `logging` module, yet it requires some care to use.
> The reason why I chose not to use it here is because our `FileLogger` in `Research` collects all the info from all the loggers and puts it into `research.log` file. Furious spam from the iteration-wise notifiers from multiple processes would make this file completely unreadable; that is why I decided to use simple file writes, provided we don't need any extra functionality (for now).

As for mixing logs with `Research` - you can just use another logger. 
As for readability: most of log files are unreadable. What you do in these logs is printing some info related to performance - this kind of data is extremely hard to understand simply by looking at a text file. You will always parse log files to build plots or to make some aggregation, so what you really need is a consistent logging format. You simply dump all data you have in a log in such a manner, that it is easy to extract what you really want to get automatically.

Reply:
> As for the not reusing code: it hurts me to say it, but `premature generalization is the root of all evil`, as one man (almost) said. The effort required to generalize similar 3 lines of code in different places may just make it harder to read and lead to more confusion, not clarity.

Well, Donald Knuth actually said nothing about evils of generalization. But the evils of duplicated code are well-known. 

The next time you need to make a change in the logging process you will have to implement it twice, and you are always at risk that you make it differently in 2 cases. Even now, are you absolutely sure that 2 different logging modes produce similar results? 

Of course you can delay the refactoring until that next time, but will not it be better to design it an a way that allows easy and reliable modifications now, that you see all context?

This particular case with logs is trivial, and it is really up to you whether to refactor it or not. But for a more important feature I would strongly insist on producing code that can be modified without pain from the very beginning.

Reply:
You are right, so I have done the refactoring requested.
As far as the logging, I still think that this class does not need the fully-fledged logger, where simple `print` to file suffices.
Also, I do love f-strings. Is there anything in the way of us updating to `Python 3.6+`, @roman-kh?

Reply:
> Also, I do love f-strings. Is there anything in the way of us updating to `Python 3.6+`, @roman-kh?
+1 for f-strings!

Reply:
Cheers for f-string.